pkg/commands/credentials.go

```diff
@@ -0,0 +1,299 @@
+type Listener int
+
+// InputQuestion is what is send by the lazygit client
+type InputQuestion struct {
+	PublicKey string
+	Question  string
+	Listener  string
+}
+
+// listenerMetaType is a listener there private key and ask function
+type listenerMetaType struct {
+	AskFunction func(string) string
+	AskedFor    struct {
+		Password bool
+		Username bool
+	}
+}
+
+var listenerMeta = map[string]listenerMetaType{} // a list of listeners
+var totalListener uint32                         // this gets used to set the key of listenerMeta
+
+// Input interacts with the lazygit client spawned by git
+func (l *Listener) Input(in InputQuestion, out *EncryptedMessage) error {
+	suspiciousErr := errors.New("closing message due to suspicious behavior")
+
+	listener, ok := listenerMeta[in.Listener]
+	if !ok {
+		return suspiciousErr
+	}
+
+	if !HasLGAsSubProcess() {
+		return suspiciousErr
+	}
+
+	updateListenerMeta := func() {
+		listenerMeta[in.Listener] = listener
+	}
+
+	updateListenerMeta()
+
+	question := in.Question
+
+	prompts := map[string]string{
+		"password": `Password\s*for\s*'.+':`,
+		"username": `Username\s*for\s*'.+':`,
+	}
+
+	var toSend string
+
+	for askFor, pattern := range prompts {
+		match, _ := regexp.MatchString(pattern, question)
+		if match && ((askFor == "password" && !listener.AskedFor.Password) || (askFor == "username" && !listener.AskedFor.Username)) {
+			switch askFor {
+			case "password":
+				listener.AskedFor.Password = true
+			case "username":
+				listener.AskedFor.Username = true
+			}
+			updateListenerMeta()
+			toSend = strings.Replace(listener.AskFunction(askFor), "\n", "", -1)
+			break
+		}
+	}
+
+	encryptedData, err := encryptMessage(in.PublicKey, toSend)
+	if err != nil {
+		return suspiciousErr
+	}
+
+	*out = encryptedData
+
+	return nil
+}
+
+// DetectUnamePass runs git commands that need credentials
+// ask() gets executed when git needs credentials
+// The ask argument will be "username" or "password"
+func (c *OSCommand) DetectUnamePass(command string, ask func(string) string) error {
+	totalListener++
+	currentListener := fmt.Sprintf("%v", totalListener)
+
+	listener := listenerMetaType{AskFunction: ask}
+	listenerMeta[currentListener] = listener
+
+	defer delete(listenerMeta, currentListener)
+
+	end := make(chan error)
+	hostPort := GetFreePort()
+	serverStartedChan := make(chan struct{})
+
+	go runGit(
+		serverStartedChan,
+		end,
+		command,
+		hostPort,
+		currentListener,
+	)
+
+	go runServer(
+		serverStartedChan,
+		end,
+		hostPort,
+		currentListener,
+	)
+
+	err := <-end
+
+	return err
+}
+
+// runGit runs the actual git command with the needed git
+func runGit(serverStartedChan chan struct{}, end chan error, command, hostPort, currentListener string) {
+	<-serverStartedChan
+
+	ex, err := os.Executable()
+	if err != nil {
+		ex = os.Args[0]
+	}
+
+	splitCmd := str.ToArgv(command)
+	cmd := exec.Command(splitCmd[0], splitCmd[1:]...)
```

Comment:
As part of the rebasing PR I recently merged, we can now take a command string and some env variables and run the command, like so:

```
	cmd := c.OSCommand.ExecutableFromString(command)
	cmd.Env = append(
		os.Environ(),
		"LAZYGIT_HOST_PORT="+hostPort,
		"LAZYGIT_LISTENER="+currentListener,

 		"GIT_ASKPASS="+ex, // tell git where lazygit is located so it can ask lazygit for credentials

 		"LANG=en_US.UTF-8",   // Force using EN as language
		"LC_ALL=en_US.UTF-8", // Force using EN as language
	)
	end <- c.OSCommand.RunExecutable(cmd)
```